File: backend/modules/petclinic/app/petclinic/api/pettype/deletePetType/DeletePetTypeService.java
package petclinic.api.pettype.deletePetType;

import akka.actor.ActorRef;
import akka.util.Timeout;
import petclinic.api.Constants;
import petclinic.api.model.PetType;
import petclinic.model.pettypes.PetTypeProtocol;
import scala.collection.JavaConverters;

import javax.inject.Inject;
import javax.inject.Named;
import java.util.Collection;
import java.util.Optional;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

public class DeletePetTypeService {

    @Inject
    @Named("pettypes")
    private ActorRef petTypesActor;


    public CompletionStage<Optional<Boolean>> deleteType(String id) {
        return akka.pattern.PatternsCS.ask(petTypesActor, PetTypeProtocol.delete(id), Constants.TIMEOUT)
                .thenApply(o -> (PetTypeProtocol.PetTypeDeleteResponse) o)
                .thenApply(o -> {
                    if (o instanceof PetTypeProtocol.PetTypeDeletedResponse$) {
                        return Optional.of(true);
                    } else if (o instanceof PetTypeProtocol.PetTypeStillInUseResponse$) {
                        return Optional.of(false);
                    } else if (o instanceof PetTypeProtocol.PetTypeToDeleteNotFoundResponse$) {
                        return Optional.empty();
                    } else {
                        throw new IllegalStateException("unhandled response received: " + o);
                    }
                });
    }
}

File: backend/modules/petclinic/app/petclinic/api/pettype/deletePetType/DeletePetTypeService.scala
package petclinic.api.pettype.deletePetType

import akka.actor.ActorRef
import javax.inject.{Inject, Named}
import petclinic.api.Constants
import petclinic.model.pettypes.PetTypeProtocol._

import scala.concurrent.{ExecutionContext, Future}

class DeletePetTypeService @Inject()(@Named(petclinic.api.Constants.PET_TYPES_ACTOR_NAME) actorRef: ActorRef)(implicit ec: ExecutionContext) {

  def deleteType(id: String): Future[Option[Boolean]] = {
    akka.pattern.Patterns.ask(actorRef, DeletePetTypeCommand(id), Constants.TIMEOUT)
      .mapTo[PetTypeDeleteResponse]
      .map {
        case PetTypeToDeleteNotFoundResponse => Some(true)
        case PetTypeDeletedResponse => None
        case PetTypeStillInUseResponse => Some(false)
      }
  }
}
